docs/decisions/0040-OTel-semantic-convention.md

```diff
@@ -0,0 +1,460 @@
+status: { proposed }
+contact: { Tao Chen }
+date: { 2024-04-15 }
+deciders: {}
+consulted: {}
+informed: {}
+---
+
+# Use standardized vocabulary and specification for observability in Semantic Kernel
+
+## Context and Problem Statement
+
+Observing LLM applications has been a huge ask from customers and the community. This work aims to ensure that SK provides the best developer experience while complying with the industry standards for observability in generative-AI-based applications.
+
+For more information, please refer to this issue: https://github.com/open-telemetry/semantic-conventions/issues/327
+
+### Semantic conventions
+
+The semantic conventions for generative AI are currently in their nascent stage, and as a result, many of the requirements outlined here may undergo changes in the future. Consequently, several features derived from this Architectural Decision Record (ADR) may be considered experimental. It is essential to remain adaptable and responsive to evolving industry standards to ensure the continuous improvement of our system's performance and reliability.
+
+- [Semantic conventions for generative AI](https://github.com/open-telemetry/semantic-conventions/tree/main/docs/gen-ai)
+- [Generic LLM attributes](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/attributes-registry/llm.md)
+
+### Telemetry requirements (Experimental)
+
+Based on the initial version, Semantic Kernel should provide the following attributes in activities that represent LLM requests:
+
+> `Activity` is a .Net concept and existed before OpenTelemetry. A `span` is an OpenTelemetry concept that is equivalent to an `Activity`.
+
+- (Required)`gen_ai.system`
+- (Required)`gen_ai.request.model`
+- (Recommended)`gen_ai.request.max_token`
+- (Recommended)`gen_ai.request.temperature`
+- (Recommended)`gen_ai.request.top_p`
+- (Recommended)`gen_ai.response.id`
+- (Recommended)`gen_ai.response.model`
+- (Recommended)`gen_ai.response.finish_reasons`
+- (Recommended)`gen_ai.response.prompt_tokens`
+- (Recommended)`gen_ai.response.completion_tokens`
+
+The following events will be optionally attached to an activity:
+| Event name| Attribute(s)|
+|---|---|
+|`gen_ai.content.prompt`|`gen_ai.prompt`|
+|`gen_ai.content.completion`|`gen_ai.completion`|
+
+> The kernel must provide configuration options to disable these events because they may contain PII.
+> See the [Semantic conventions for generative AI](https://github.com/open-telemetry/semantic-conventions/tree/main/docs/gen-ai) for requirement level for these attributes.
+
+## Where do we create the activities
+
+It is crucial to establish a clear line of responsibilities, particularly since certain service providers, such as the Azure OpenAI SDK, have pre-existing instrumentation (in which case SK won't need to track LLM requests). Our objective is to position our activities as close to the model level as possible to promote a more cohesive and consistent developer experience.
```

Comment:
The Azure SDK team is planning to adapt to the OTel semantic conventions for LLMs.